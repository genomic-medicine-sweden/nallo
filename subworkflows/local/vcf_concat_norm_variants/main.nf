include { ADD_FOUND_IN_TAG                           } from '../../../modules/local/add_found_in_tag/main'
include { BCFTOOLS_CONCAT                             } from '../../../modules/nf-core/bcftools/concat/main'
include { BCFTOOLS_NORM as BCFTOOLS_NORM_SINGLESAMPLE } from '../../../modules/nf-core/bcftools/norm/main'

//
// Workflow to concatenate and normalize variants
//
workflow VCF_CONCAT_NORM_VARIANTS {
    take:
    ch_vcfs        // channel: [mandatory] [Â val(meta), path(vcf) ]
    ch_fasta       // channel: [mandatory] [ val(meta), path(fasta) ]
    variant_caller // string: variant caller to tag the variants with, e.g. "deepvariant"

    main:
    ch_versions = Channel.empty()

    BCFTOOLS_CONCAT(
        ch_vcfs.map { meta, vcfs -> [meta, vcfs, []] },
    )
    ch_versions = ch_versions.mix(BCFTOOLS_CONCAT.out.versions)

    // Annotate with FOUND_IN tag - not sure what would happen if we do this before glnexus instead?
    ADD_FOUND_IN_TAG(
        BCFTOOLS_CONCAT.out.vcf.map { meta, vcf -> [meta, vcf, []] },
        variant_caller,
    )
    ch_versions = ch_versions.mix(ADD_FOUND_IN_TAG.out.versions)

    BCFTOOLS_NORM_SINGLESAMPLE(
        ADD_FOUND_IN_TAG.out.vcf.map { meta, vcf -> [meta, vcf, []] },
        ch_fasta,
    )
    ch_versions = ch_versions.mix(BCFTOOLS_NORM_SINGLESAMPLE.out.versions)

    emit:
    vcf                 = BCFTOOLS_NORM_SINGLESAMPLE.out.vcf                                         // channel: [ val(meta), path(vcf) ]
    index               = BCFTOOLS_NORM_SINGLESAMPLE.out.tbi.mix(BCFTOOLS_NORM_SINGLESAMPLE.out.csi) // channel: [ val(meta), path(tbi/csi) ]
    bcftools_concat_vcf = BCFTOOLS_CONCAT.out.vcf                                                    // channel: [ val(meta), path(vcf) ]
    versions            = ch_versions                                                                // channel: [ path(versions.yml) ]
}
