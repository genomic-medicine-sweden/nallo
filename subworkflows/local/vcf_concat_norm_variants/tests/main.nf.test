nextflow_workflow {

    name "Test Workflow VCF_CONCAT_NORM_VARIANTS"
    script "../main.nf"
    config "./nextflow.config"
    workflow "VCF_CONCAT_NORM_VARIANTS"

    setup {
        run("GUNZIP") {
            script "../../../../modules/nf-core/gunzip/main.nf"
            process {
                """
                input[0] = [
                    [ id:'test' ],
                    file(params.pipelines_testdata_base_path + 'reference/hg38.test.fa.gz', checkIfExists: true)
                ]
                """
            }

        }
    }

    test("1 VCF") {

        when {
            workflow {
                """
                input[0] = Channel.of([
                    [ id: 'test' ],
                    file(params.pipelines_testdata_base_path + 'testdata/HG002_deepvariant_norm_chr20.vcf.gz', checkIfExists: true),
                ])
                input[1] = GUNZIP.out.gunzip
                input[2] = 'deepvariant'
                """
            }
        }

        then {
            assertAll(
                { assert workflow.success },
                { assert snapshot(
                    workflow.out,
                ).match() }
            )
        }

    }

    test("2 VCFs") {

        when {
            workflow {
                """
                input[0] = Channel.of([
                    [ id: 'test' ],
                    [
                        file(params.pipelines_testdata_base_path + 'testdata/HG002_deepvariant_norm_chr20.vcf.gz', checkIfExists: true),
                        file(params.pipelines_testdata_base_path + 'testdata/HG002_deepvariant_norm_chrX.vcf.gz', checkIfExists: true),
                    ]
                ])
                input[1] = GUNZIP.out.gunzip
                input[2] = 'deepvariant'
                """
            }
        }

        then {
            assertAll(
                { assert workflow.success },
                { assert snapshot(
                    workflow.out,
                ).match() }
            )
        }

    }

    test("1 VCF -stub") {

        when {
            workflow {
                """
                input[0] = Channel.of([
                    [ id: 'test' ],
                    file(params.pipelines_testdata_base_path + 'testdata/HG002_deepvariant_norm_chr20.vcf.gz', checkIfExists: true),
                ])
                input[1] = GUNZIP.out.gunzip
                input[2] = 'deepvariant'
                """
            }
        }

        then {
            assertAll(
                { assert workflow.success },
                { assert snapshot(
                    workflow.out,
                ).match() }
            )
        }

    }

    test("2 VCFs -stub") {

        when {
            workflow {
                """
                input[0] = Channel.of([
                    [ id: 'test' ],
                    [
                        file(params.pipelines_testdata_base_path + 'testdata/HG002_deepvariant_norm_chr20.vcf.gz', checkIfExists: true),
                        file(params.pipelines_testdata_base_path + 'testdata/HG002_deepvariant_norm_chrX.vcf.gz', checkIfExists: true),
                    ]
                ])
                input[1] = GUNZIP.out.gunzip
                input[2] = 'deepvariant'
                """
            }
        }

        then {
            assertAll(
                { assert workflow.success },
                { assert snapshot(
                    workflow.out,
                ).match() }
            )
        }

    }
}
