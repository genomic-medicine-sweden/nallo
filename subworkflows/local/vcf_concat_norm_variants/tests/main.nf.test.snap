{
    "2 VCFs -stub": {
        "content": [
            {
                "0": [
                    [
                        {
                            "id": "test"
                        },
                        "test_deepvariant_snvs.vcf.gz:md5,a05f86f77628989fb2870e82e9276474"
                    ]
                ],
                "1": [
                    [
                        {
                            "id": "test"
                        },
                        "test_deepvariant_snvs.vcf.gz.tbi:md5,bb95c6bbb767007b3598eab786233c86"
                    ]
                ],
                "2": [
                    [
                        {
                            "id": "test"
                        },
                        "test.vcf:md5,1b73defba4121ae2e70c49db5de5172f"
                    ]
                ],
                "3": [
                    "versions.yml:md5,2678b6183e1782cb60da7a21e359f38d",
                    "versions.yml:md5,44e9c6136408769fdeba5aec27612518",
                    "versions.yml:md5,7f07c70c14670d90338aaa59cdbf780c"
                ],
                "bcftools_concat_vcf": [
                    [
                        {
                            "id": "test"
                        },
                        "test.vcf:md5,1b73defba4121ae2e70c49db5de5172f"
                    ]
                ],
                "index": [
                    [
                        {
                            "id": "test"
                        },
                        "test_deepvariant_snvs.vcf.gz.tbi:md5,bb95c6bbb767007b3598eab786233c86"
                    ]
                ],
                "vcf": [
                    [
                        {
                            "id": "test"
                        },
                        "test_deepvariant_snvs.vcf.gz:md5,a05f86f77628989fb2870e82e9276474"
                    ]
                ],
                "versions": [
                    "versions.yml:md5,2678b6183e1782cb60da7a21e359f38d",
                    "versions.yml:md5,44e9c6136408769fdeba5aec27612518",
                    "versions.yml:md5,7f07c70c14670d90338aaa59cdbf780c"
                ]
            }
        ],
        "meta": {
            "nf-test": "0.9.2",
            "nextflow": "25.04.6"
        },
        "timestamp": "2025-10-22T11:12:20.468384016"
    },
    "2 VCFs": {
        "content": [
            {
                "0": [
                    [
                        {
                            "id": "test"
                        },
                        "test_deepvariant_snvs.vcf.gz:md5,a05f86f77628989fb2870e82e9276474"
                    ]
                ],
                "1": [
                    [
                        {
                            "id": "test"
                        },
                        "test_deepvariant_snvs.vcf.gz.tbi:md5,bb95c6bbb767007b3598eab786233c86"
                    ]
                ],
                "2": [
                    [
                        {
                            "id": "test"
                        },
                        "test.vcf:md5,1b73defba4121ae2e70c49db5de5172f"
                    ]
                ],
                "3": [
                    "versions.yml:md5,2678b6183e1782cb60da7a21e359f38d",
                    "versions.yml:md5,44e9c6136408769fdeba5aec27612518",
                    "versions.yml:md5,7f07c70c14670d90338aaa59cdbf780c"
                ],
                "bcftools_concat_vcf": [
                    [
                        {
                            "id": "test"
                        },
                        "test.vcf:md5,1b73defba4121ae2e70c49db5de5172f"
                    ]
                ],
                "index": [
                    [
                        {
                            "id": "test"
                        },
                        "test_deepvariant_snvs.vcf.gz.tbi:md5,bb95c6bbb767007b3598eab786233c86"
                    ]
                ],
                "vcf": [
                    [
                        {
                            "id": "test"
                        },
                        "test_deepvariant_snvs.vcf.gz:md5,a05f86f77628989fb2870e82e9276474"
                    ]
                ],
                "versions": [
                    "versions.yml:md5,2678b6183e1782cb60da7a21e359f38d",
                    "versions.yml:md5,44e9c6136408769fdeba5aec27612518",
                    "versions.yml:md5,7f07c70c14670d90338aaa59cdbf780c"
                ]
            }
        ],
        "meta": {
            "nf-test": "0.9.2",
            "nextflow": "25.04.6"
        },
        "timestamp": "2025-10-22T11:12:06.822154392"
    },
    "1 VCF": {
        "content": [
            {
                "0": [
                    [
                        {
                            "id": "test"
                        },
                        "test_deepvariant_snvs.vcf.gz:md5,ce1219f6e40f0a819838772321f9237d"
                    ]
                ],
                "1": [
                    [
                        {
                            "id": "test"
                        },
                        "test_deepvariant_snvs.vcf.gz.tbi:md5,96078fdf66a8c75b797c2a22ac539c8c"
                    ]
                ],
                "2": [
                    [
                        {
                            "id": "test"
                        },
                        "test.vcf:md5,266dba05c621383d4c7ebc3cc1b8fa41"
                    ]
                ],
                "3": [
                    "versions.yml:md5,2678b6183e1782cb60da7a21e359f38d",
                    "versions.yml:md5,44e9c6136408769fdeba5aec27612518",
                    "versions.yml:md5,7f07c70c14670d90338aaa59cdbf780c"
                ],
                "bcftools_concat_vcf": [
                    [
                        {
                            "id": "test"
                        },
                        "test.vcf:md5,266dba05c621383d4c7ebc3cc1b8fa41"
                    ]
                ],
                "index": [
                    [
                        {
                            "id": "test"
                        },
                        "test_deepvariant_snvs.vcf.gz.tbi:md5,96078fdf66a8c75b797c2a22ac539c8c"
                    ]
                ],
                "vcf": [
                    [
                        {
                            "id": "test"
                        },
                        "test_deepvariant_snvs.vcf.gz:md5,ce1219f6e40f0a819838772321f9237d"
                    ]
                ],
                "versions": [
                    "versions.yml:md5,2678b6183e1782cb60da7a21e359f38d",
                    "versions.yml:md5,44e9c6136408769fdeba5aec27612518",
                    "versions.yml:md5,7f07c70c14670d90338aaa59cdbf780c"
                ]
            }
        ],
        "meta": {
            "nf-test": "0.9.2",
            "nextflow": "25.04.6"
        },
        "timestamp": "2025-10-22T11:11:59.928419495"
    },
    "1 VCF -stub": {
        "content": [
            {
                "0": [
                    [
                        {
                            "id": "test"
                        },
                        "test_deepvariant_snvs.vcf.gz:md5,ce1219f6e40f0a819838772321f9237d"
                    ]
                ],
                "1": [
                    [
                        {
                            "id": "test"
                        },
                        "test_deepvariant_snvs.vcf.gz.tbi:md5,96078fdf66a8c75b797c2a22ac539c8c"
                    ]
                ],
                "2": [
                    [
                        {
                            "id": "test"
                        },
                        "test.vcf:md5,266dba05c621383d4c7ebc3cc1b8fa41"
                    ]
                ],
                "3": [
                    "versions.yml:md5,2678b6183e1782cb60da7a21e359f38d",
                    "versions.yml:md5,44e9c6136408769fdeba5aec27612518",
                    "versions.yml:md5,7f07c70c14670d90338aaa59cdbf780c"
                ],
                "bcftools_concat_vcf": [
                    [
                        {
                            "id": "test"
                        },
                        "test.vcf:md5,266dba05c621383d4c7ebc3cc1b8fa41"
                    ]
                ],
                "index": [
                    [
                        {
                            "id": "test"
                        },
                        "test_deepvariant_snvs.vcf.gz.tbi:md5,96078fdf66a8c75b797c2a22ac539c8c"
                    ]
                ],
                "vcf": [
                    [
                        {
                            "id": "test"
                        },
                        "test_deepvariant_snvs.vcf.gz:md5,ce1219f6e40f0a819838772321f9237d"
                    ]
                ],
                "versions": [
                    "versions.yml:md5,2678b6183e1782cb60da7a21e359f38d",
                    "versions.yml:md5,44e9c6136408769fdeba5aec27612518",
                    "versions.yml:md5,7f07c70c14670d90338aaa59cdbf780c"
                ]
            }
        ],
        "meta": {
            "nf-test": "0.9.2",
            "nextflow": "25.04.6"
        },
        "timestamp": "2025-10-22T11:12:13.762657632"
    }
}